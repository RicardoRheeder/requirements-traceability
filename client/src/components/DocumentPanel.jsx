import React, { useEffect, useState } from 'react'
import Dropdown from 'react-dropdown'
import { useHistory } from 'react-router-dom'

import { useDispatch, useSelector } from 'react-redux'
import {
  setSelectedDocumentPanelObject,
  setModalObject,
  updateDataTree,
  setCurrentDocVersion,
} from '../redux/stores/common/actions'
import {
  getDocAsync,
  updateCurrentDocument,
} from '../redux/stores/document/actions'

export const DocumentPanel = ({ document }) => {
  const dispatch = useDispatch()
  const history = useHistory()

  const selectedDocumentPanelObject = useSelector(
    (state) => state.common.selectedDocumentPanelObject
  )

  const [selectedVersionTree, setSelectedVersionTree] = useState()
  const [versionList, setVersionList] = useState([])
  const [currentDropDownVersion, setCurrentDropDownVersion] = useState('')

  useEffect(() => {
    refreshVersionList()
  }, [])

  function refreshVersionList() {
    let defaultOption = '0.0'
    let tempVersionsList = []
    if (document.versions.length > 0) {
      // looping over versions array and parsing
      document.versions.forEach((version) => {
        const parsedVersion = JSON.parse(version)
        tempVersionsList.push(parsedVersion.versionName)
      })
      tempVersionsList.reverse()
      setSelectedVersionTree(JSON.parse(document.tree))
      // setting default option
      defaultOption = tempVersionsList[0]
      setCurrentDropDownVersion(defaultOption)
      setVersionList(tempVersionsList)
    } else {
      setCurrentDropDownVersion(defaultOption)
      setSelectedVersionTree(JSON.parse(document.tree))
    }
  }

  const _onDropdownSelect = (thing) => {
    // finding the corresponding tree for the version that was selected
    document.versions.forEach((version) => {
      const parsedVersion = JSON.parse(version)

      if (thing.value == parsedVersion.versionName) {
        dispatch(updateDataTree(JSON.parse(parsedVersion.tree)))
        setCurrentDropDownVersion(parsedVersion.versionName)
        setSelectedVersionTree(JSON.parse(parsedVersion.tree))
      }
    })
  }

  const inviteUserButton = () => {
    dispatch(setModalObject({ visible: true, mode: 2 }))
  }

  const openDocumentIntoEditor = () => {
    // fetching the current document
    dispatch(getDocAsync(document._id))
    // updating the selected version
    dispatch(updateDataTree(selectedVersionTree))
    // setting the current version of the document
    dispatch(setCurrentDocVersion(currentDropDownVersion))
    history.push('/editor')
  }

  return (
    <div
      className={
        'document-panel-component' +
        (selectedDocumentPanelObject._id == document._id ? ' selected' : '')
      }
      onClick={() => {
        dispatch(setSelectedDocumentPanelObject(document))
      }}
      onDoubleClick={openDocumentIntoEditor}
    >
      <div className="document-panel-title">
        <button className="add-person-button" onClick={inviteUserButton}>
          <img
            className="add-person-button-image"
            src="./assets/images/add-friend-icon.png"
          ></img>
        </button>
        <h2 className="doc-title">{document.title}</h2>
      </div>

      <div className="document-panel-dropdown">
        {/* <Dropdown
          options={versionList}
          onChange={_onDropdownSelect}
          value={currentDropDownVersion}
          placeholder="Select an option"
          className="dropdown-custom-wrapper"
          controlClassName="dropdown-custom-control"
          placeholderClassName="dropdown-custom-placeholder"
          menuClassName="dropdown-custom-menu"
          arrowClassName="dropdown-custom-arrow"
        /> */}
      </div>
    </div>
  )
}
