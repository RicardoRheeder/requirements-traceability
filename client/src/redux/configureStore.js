import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

// restricting extension in production using process.env.NODE_ENV which is set 'production'
import { composeWithDevTools } from 'redux-devtools-extension/logOnlyInProduction';

import store from './stores';

const enhancer = (process.env.NODE_ENV == 'development')
               ? composeWithDevTools(applyMiddleware(thunk))
               : applyMiddleware(thunk);

export default () => createStore(store, enhancer);
