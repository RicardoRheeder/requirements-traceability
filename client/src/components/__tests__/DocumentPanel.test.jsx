import React from "react"
import Enzyme, { shallow, mount} from "enzyme"
import {DocumentPanel} from "../DocumentPanel"
import Adapter from "enzyme-adapter-react-16"
import { Provider } from "react-redux";
import configureStore from "redux-mock-store"


const mockStore = configureStore([])
Enzyme.configure({adapter: new Adapter() });

describe("Document panel", () => {
    let store;

    beforeEach(() => {
        store = mockStore({});
    });
    test("renders", () => {
        const wrapper = shallow(<Provider store = {store}><DocumentPanel/></Provider>);
        expect(wrapper.exists()).toBe(true);
        expect(wrapper).toMatchSnapshot();
    });
});