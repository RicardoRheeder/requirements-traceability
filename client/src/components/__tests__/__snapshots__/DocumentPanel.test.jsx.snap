// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Document panel renders 1`] = `ShallowWrapper {}`;
